waku/v2/node/builder.nim

```diff
@@ -0,0 +1,167 @@
+when (NimMajor, NimMinor) < (1, 4):
+  {.push raises: [Defect].}
+else:
+  {.push raises: [].}
+
+import
+  std/options,
+  stew/results,
+  stew/shims/net,
+  chronicles,
+  libp2p/crypto/crypto,
+  libp2p/builders,
+  libp2p/nameresolving/nameresolver
+import
+  ../protocol/waku_enr,
+  ../protocol/waku_discv5,
+  ./config,
+  ./peer_manager,
+  ./waku_node
+
+
+type
+  WakuNodeBuilder* = object
+    # General
+    nodeRng: Option[ref crypto.HmacDrbgContext]
+    nodeKey: Option[crypto.PrivateKey]
+    netConfig: Option[NetConfig]
+
+    # Peer storage and peer manager
+    peerStorage: Option[PeerStorage]
+    peerStorageCapacity: Option[int]
+    peerManager: Option[PeerManager]
+
+    # Libp2p switch
+    switchMaxConnections: Option[int]
+    switchNameResolver: Option[NameResolver]
+    switchAgentString: Option[string]
+    switchSslSecureKey: Option[string]
+    switchSslSecureCert: Option[string]
+    switchSendSignedPeerRecord: Option[bool]
+
+    # Waku discv5
+    wakuDiscv5: Option[WakuDiscoveryV5]
+
+  WakuNodeBuilderResult* = Result[void, string]
+
+
+## Init
+
+proc init*(T: type WakuNodeBuilder): WakuNodeBuilder =
+  WakuNodeBuilder()
+
+
+## General
+
+proc withRng*(builder: var WakuNodeBuilder, rng: ref crypto.HmacDrbgContext) =
+  builder.nodeRng = some(rng)
+
+proc withNodeKey*(builder: var WakuNodeBuilder, nodeKey: crypto.PrivateKey) =
+  builder.nodeKey = some(nodeKey)
+
+proc withNetworkConfiguration*(builder: var WakuNodeBuilder, config: NetConfig) =
+  builder.netConfig = some(config)
+
+proc withNetworkConfigurationDetails*(builder: var WakuNodeBuilder,
+          bindIp: ValidIpAddress,
+          bindPort: Port,
+          extIp = none(ValidIpAddress),
+          extPort = none(Port),
+          extMultiAddrs = newSeq[MultiAddress](),
+          wsBindPort: Port = Port(8000),
+          wsEnabled: bool = false,
+          wssEnabled: bool = false,
+          wakuFlags = none(CapabilitiesBitfield),
+          dns4DomainName = none(string),
+          discv5UdpPort = none(Port)): WakuNodeBuilderResult {.
+  deprecated: "use 'builder.withNetworkConfiguration()' instead".} =
+  let netConfig = ? NetConfig.init(
+    bindIp = bindIp,
+    bindPort = bindPort,
+    extIp = extIp,
+    extPort = extPort,
+    extMultiAddrs = extMultiAddrs,
+    wsBindPort = wsBindPort,
+    wsEnabled = wsEnabled,
+    wssEnabled = wssEnabled,
+    wakuFlags = wakuFlags,
+    dns4DomainName = dns4DomainName,
+    discv5UdpPort = discv5UdpPort,
+  )
+  builder.withNetworkConfiguration(netConfig)
+  ok()
+
+
+## Peer storage and peer manager
+
+proc withPeerStorage*(builder: var WakuNodeBuilder, peerStorage: PeerStorage, capacity = none(int)) =
```

Comment:
perhaps merge with withPeerManager?

Reply:
Yes! This is the idea. It will be merged after we can change the `WakuNode.new()` proc without having to change 120 test cases :)